qiskit/pulse/instruction_schedule_map.py

```diff
@@ -0,0 +1,293 @@
+            schedule: [Schedule, ParameterizedSchedule]) -> None:
+        """
+        Add a new known instruction.
+
+        Args:
+            instruction: The name of the instruction to add.
+            qubits: The qubits which the instruction applies to.
+            schedule: The Schedule that implements the given instruction.
+
+        Returns:
+            None
+
+        Raises:
+            PulseError: If the qubits are provided as an empty iterable.
+        """
+        qubits = _to_tuple(qubits)
+        if qubits == ():
+            raise PulseError("Cannot add definition {} with no target qubits.".format(instruction))
+        if not isinstance(schedule, (Schedule, ParameterizedSchedule)):
+            raise PulseError("Attemping to add an invalid schedule type.")
+        self._map[instruction][qubits] = schedule
+        self._qubit_insts[qubits].add(instruction)
+
+    def remove(self, instruction: str, qubits: Union[int, Iterable[int]]) -> None:
+        """Remove the given instruction from the defined instructions.
+
+        Args:
+            instruction: The name of the instruction to add.
+            qubits: The qubits which the instruction applies to.
+
+        Returns:
+            None
+        """
+        qubits = _to_tuple(qubits)
+        self.assert_has(instruction, qubits)
+        self._map[instruction].pop(qubits)
+        self._qubit_insts[qubits].remove(instruction)
+        if not self._map[instruction]:
+            self._map.pop(instruction)
+        if not self._qubit_insts[qubits]:
+            self._qubit_insts.pop(qubits)
+
+    def pop(self,
+            instruction: str,
+            qubits: Union[int, Iterable[int]],
+            *params: List[Union[int, float, complex]],
+            **kwparams: Dict[str, Union[int, float, complex]]) -> Schedule:
+        """
+        Remove and return the defined Schedule for the given instruction on the given qubits.
+
+        Args:
+            instruction: Name of the instruction.
+            qubits: The qubits for the instruction.
+            *params: Command parameters for generating the output schedule.
+            **kwparams: Keyworded command parameters for generating the schedule.
+
+        Returns:
+            The Schedule defined for the input.
+        """
+        self.assert_has(instruction, qubits)
+        schedule = self._map[instruction][_to_tuple(qubits)]
+        if isinstance(schedule, ParameterizedSchedule):
+            return schedule.bind_parameters(*params, **kwparams)
+        self.remove(instruction, qubits)
+        return schedule
+
+    def cmds(self) -> List[str]:
+        """
+        Deprecated.
+
+        Returns:
+            The names of all the circuit instructions which have Schedule definitions in this.
+        """
+        warnings.warn("Please use the `instructions` attribute instead of `cmds()`.",
+                      DeprecationWarning)
+        return copy.copy(self.instructions)
```

Comment:
@taalexander why did you make this change? `self.instructions` is generated fresh each time